apps/cpp_rpc/rpc_env.cc

```diff
@@ -0,0 +1,247 @@
+        }
+        f += d->d_name;
+        vec.push_back(f);
+      }
+    }
+    closedir(dp);
+  #else
+    WIN32_FIND_DATA fd;
+    std::string pattern = dirname + "/*";
+    HANDLE handle = FindFirstFile(pattern.c_str(), &fd);
+    if (handle == INVALID_HANDLE_VALUE) {
+      int errsv = GetLastError();
+      LOG(FATAL) << "ListDir " << dirname << " error: " << strerror(errsv);
+    }
+    do {
+      if (fd.cFileName != "." && fd.cFileName != "..") {
+        std::string  f = dirname;
+        char clast = f[f.length() - 1];
+        if (f == ".") {
+          f = fd.cFileName;
+        } else if (clast != '/' && clast != '\\') {
+          f += '/';
+          f += fd.cFileName;
+        }
+        vec.push_back(f);
+      }
+    }  while (FindNextFile(handle, &fd));
+    FindClose(handle);
+  #endif
+  return vec;
+}
+
+/*!
+ * \brief LinuxShared Creates a linux shared library
+ * \param output The output file name
+ * \param files The files for building
+ * \param options The compiler options
+ * \param cc The compiler
+ */
+void LinuxShared(const std::string output,
+                 const std::vector<std::string> &files,
+                 std::string options = "",
+                 std::string cc = "g++") {
+    std::string cmd = cc;
+    cmd += " -shared -fPIC ";
+    cmd += " -o " + output;
+    for (auto f = files.begin(); f != files.end(); ++f) {
+     cmd += " " + *f;
+    }
+    cmd += " " + options;
+    CHECK(system(cmd.c_str()) == 0) << "Compilation error.";
+}
+
+/*!
+ * \brief CreateShared Creates a shared library
+ * \param output The output file name
+ * \param files The files for building
+ */
+void CreateShared(const std::string output, const std::vector<std::string> &files) {
+  #if defined(__linux__) || defined(__ANDROID__)
+    LinuxShared(output, files);
+  #else
+    LOG(FATAL) << "Do not support creating shared library";
+  #endif
+}
+
+/*!
+ * \brief Load Load module from file
+          This function will automatically call
+          cc.create_shared if the path is in format .o or .tar
+          High level handling for .o and .tar file.
+          We support this to be consistent with RPC module load.
+ * \param fileIn The input file, file name will be updated
+ * \param fmt The format of file
+ * \return Module The loaded module
+ */
+Module Load(std::string *fileIn, const std::string fmt) {
+  std::string file = *fileIn;
+  if (common::EndsWith(file, ".so")) {
+      return Module::LoadFromFile(file, fmt);
+  }
+
+  #if defined(__linux__) || defined(__ANDROID__)
+    std::string file_name = file + ".so";
+    if (common::EndsWith(file, ".o")) {
+      std::vector<std::string> files;
+      files.push_back(file);
+      CreateShared(file_name, files);
+    } else if (common::EndsWith(file, ".tar")) {
+      std::string tmp_dir = "./rpc/tmp/";
+      mkdir(&tmp_dir[0], 0777);
+      std::string cmd = "tar -C " + tmp_dir + " -zxf " + file;
+      CHECK(system(cmd.c_str()) == 0) << "Untar library error.";
```

Comment:
It would be great to have a wrapper similar to python's popen: where we can capture the stderr using pipe and and throw it in the error message

Reply:
```c++

We can support something like the following code, this will allow us to throw the error message in the stdout to make the error more meaningful.

int POpen(std::string cmd, std::string* stdout) {
    // call into popen or window's popen to capture the error and throw if there is an error.
    // redirect stderr to stdout, can also be used on the caller side.
    cmd += " 2>&1";
    std::string stdout = GetStdOutFromFileUnltilEOF(fp);
    status = pclose(fp);
    if (WIFEXITED(status))  return WEXITSTATUS(status);
    return 255;
}
```

Reply:
@tqchen How about this? https://github.com/apache/incubator-tvm/pull/4281/files#diff-1853aa49a938f344ec71a548f8353dcbR108-R117

It will be check the result length and output the error result string.

Reply:
You will need to check the status returned by PClose. Otherwise the target command could print something(e.g. warning or just logs) but exit with 0 status, which is still valid.

Reply:
Done.